File: src/controllers/borrowController.js
const Book = require("../models/Book");
const BorrowRecord = require("../models/BorrowRecord");
const User = require("../models/User");
const {
  BORROW_STATUS,
  MEMBERSHIP_STATUS,
  NOTIFICATION_TYPES,
  ROLES
} = require("../constants/roles");
const catchAsync = require("../utils/catchAsync");
const AppError = require("../utils/AppError");
const { calculateFine } = require("../utils/fines");
const { createNotification } = require("../utils/notifications");

const getDefaultDueDate = () => {
  const borrowDays = Number(process.env.BORROW_DAYS || 14);
  const dueDate = new Date();
  dueDate.setDate(dueDate.getDate() + borrowDays);
  return dueDate;
};

const markOverdueRecords = async () => {
  const overdueRecords = await BorrowRecord.find({
    status: BORROW_STATUS.BORROWED,
    dueDate: { $lt: new Date() }
  }).populate("book", "title");

  await Promise.all(overdueRecords.map(async (record) => {
    record.status = BORROW_STATUS.OVERDUE;
    await record.save({ validateModifiedOnly: true });

    await createNotification({
      user: record.member,
      type: NOTIFICATION_TYPES.OVERDUE_ALERT,
      title: "Book overdue",
      message: `The book "${record.book.title}" is overdue.`,
      relatedBorrowRecord: record._id,
      dueDate: record.dueDate
    });
  }));
};

const resolveBorrowingMember = async (req) => {
  if (req.user.role === ROLES.LIBRARIAN) {
    if (!req.body.memberId) {
      throw new AppError("memberId is required when a librarian issues a book.", 400);
    }

    return User.findOne({
      _id: req.body.memberId,
      role: ROLES.MEMBER
    });
  }

  if (req.body.memberId && req.body.memberId.toString() !== req.user._id.toString()) {
    throw new AppError("Members can only borrow books for themselves.", 403);
  }

  return req.user;
};

const borrowBook = catchAsync(async (req, res, next) => {
  const member = await resolveBorrowingMember(req);

  if (!member) {
    return next(new AppError("Member not found.", 404));
  }

  if (member.membershipStatus !== MEMBERSHIP_STATUS.ACTIVE) {
    return next(new AppError("Only active members can borrow books.", 403));
  }

  const activeBorrowCount = await BorrowRecord.countDocuments({
    member: member._id,
    status: { $in: [BORROW_STATUS.BORROWED, BORROW_STATUS.OVERDUE] }
  });

  if (activeBorrowCount >= member.borrowLimit) {
    return next(new AppError("Borrow limit reached for this member.", 400));
  }

  const book = await Book.findById(req.params.bookId);

  if (!book) {
    return next(new AppError("Book not found.", 404));
  }

  if (book.availableCopies < 1) {
    return next(new AppError("This book is currently unavailable.", 400));
  }

  const dueDate = req.body.dueDate ? new Date(req.body.dueDate) : getDefaultDueDate();

  const record = await BorrowRecord.create({
    member: member._id,
    book: book._id,
    issuedBy: req.user._id,
    dueDate,
    notes: req.body.notes
  });

  book.availableCopies -= 1;
  await book.save({ validateModifiedOnly: true });

  await createNotification({
    user: member._id,
    type: NOTIFICATION_TYPES.BORROW_CONFIRMATION,
    title: "Book borrowed",
    message: `You borrowed "${book.title}". Due date: ${dueDate.toDateString()}.`,
    relatedBorrowRecord: record._id,
    dueDate
  });

  const populatedRecord = await BorrowRecord.findById(record._id)
    .populate("member", "name email")
    .populate("book", "title isbn author")
    .populate("issuedBy", "name email");

  return res.status(201).json({
    status: "success",
    data: {
      borrowRecord: populatedRecord
    }
  });
});

const returnBook = catchAsync(async (req, res, next) => {
  const record = await BorrowRecord.findById(req.params.recordId)
    .populate("book")
    .populate("member", "name email");

  if (!record) {
    return next(new AppError("Borrow record not found.", 404));
  }

  if (
    req.user.role === ROLES.MEMBER &&
    record.member._id.toString() !== req.user._id.toString()
  ) {
    return next(new AppError("Members can only return their own borrowed books.", 403));
  }

  if (record.status === BORROW_STATUS.RETURNED) {
    return next(new AppError("This book has already been returned.", 400));
  }

  const returnedAt = new Date();
  record.status = BORROW_STATUS.RETURNED;
  record.returnDate = returnedAt;
  record.returnedTo = req.user.role === ROLES.LIBRARIAN ? req.user._id : undefined;
  record.fine = calculateFine(record.dueDate, returnedAt);
  record.notes = req.body.notes || record.notes;
  await record.save({ validateModifiedOnly: true });

  record.book.availableCopies += 1;
  await record.book.save({ validateModifiedOnly: true });

  await createNotification({
    user: record.member._id,
    type: NOTIFICATION_TYPES.RETURN_CONFIRMATION,
    title: "Book returned",
    message: `You returned "${record.book.title}". Fine: ${record.fine}.`,
    relatedBorrowRecord: record._id
  });

  return res.status(200).json({
    status: "success",
    data: {
      borrowRecord: record
    }
  });
});

const getBorrowingHistory = catchAsync(async (req, res) => {
  const {
    page,
    limit,
    sort,
    status
  } = req.query;

  const filter = {};

  if (req.user.role === ROLES.MEMBER) {
    filter.member = req.user._id;
  }

  if (status) {
    filter.status = status;
  }

  const skip = (page - 1) * limit;
  const [records, total] = await Promise.all([
    BorrowRecord.find(filter)
      .populate("member", "name email")
      .populate("book", "title isbn author")
      .populate("issuedBy", "name email")
      .populate("returnedTo", "name email")
      .sort(sort)
      .skip(skip)
      .limit(limit),
    BorrowRecord.countDocuments(filter)
  ]);

  res.status(200).json({
    status: "success",
    results: records.length,
    pagination: {
      page,
      limit,
      total,
      pages: Math.ceil(total / limit)
    },
    data: {
      records
    }
  });
});

const getOverdueRecords = catchAsync(async (req, res) => {
  await markOverdueRecords();

  const records = await BorrowRecord.find({
    status: BORROW_STATUS.OVERDUE
  })
    .populate("member", "name email phone")
    .populate("book", "title isbn author")
    .sort("dueDate");

  res.status(200).json({
    status: "success",
    results: records.length,
    data: {
      records
    }
  });
});

const checkAvailability = catchAsync(async (req, res, next) => {
  const book = await Book.findById(req.params.bookId).populate("category", "name");

  if (!book) {
    return next(new AppError("Book not found.", 404));
  }

  return res.status(200).json({
    status: "success",
    data: {
      bookId: book._id,
      title: book.title,
      totalCopies: book.totalCopies,
      availableCopies: book.availableCopies,
      isAvailable: book.availableCopies > 0
    }
  });
});

module.exports = {
  borrowBook,
  returnBook,
  getBorrowingHistory,
  getOverdueRecords,
  checkAvailability
};

File: java-backend/BorrowController.java
import com.sun.net.httpserver.HttpExchange;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.UUID;
import java.util.stream.Collectors;

class BorrowController {
  private final LibraryDataStore store;
  private final SecurityService security;
  private final JsonViews views;
  private final BorrowService borrowService;

  BorrowController(LibraryDataStore store, SecurityService security, JsonViews views, BorrowService borrowService) {
    this.store = store;
    this.security = security;
    this.views = views;
    this.borrowService = borrowService;
  }

  void borrow(HttpExchange exchange, String bookId) throws IOException {
    User user = security.requireAuth(exchange);
    if (user == null) {
      return;
    }

    String requestBody = RequestUtil.body(exchange);
    User member = "librarian".equals(user.role)
      ? resolveIssueMember(exchange, requestBody)
      : user;
    Book book = store.findBook(bookId);

    if (member == null) {
      return;
    }

    if (member == null || !"member".equals(member.role)) {
      HttpResponses.sendError(exchange, 404, "Member not found.");
      return;
    }
    if (!"active".equals(member.membershipStatus)) {
      HttpResponses.sendError(exchange, 403, "Only active members can borrow books.");
      return;
    }
    if (book == null) {
      HttpResponses.sendError(exchange, 404, "Book not found.");
      return;
    }
    if (book.availableCopies < 1) {
      HttpResponses.sendError(exchange, 400, "This book is currently unavailable.");
      return;
    }

    BorrowRecord record = new BorrowRecord("borrow-" + UUID.randomUUID(), member.id, book.id, user.id, "borrowed", TimeUtil.now(), TimeUtil.daysFromNow(14), null, TimeUtil.now());
    record.notes = JsonUtil.jsonString(requestBody, "notes", "");
    store.borrowRecords().add(record);
    book.availableCopies -= 1;
    book.updatedAt = TimeUtil.now();
    store.notifications().add(new Notification("notification-" + UUID.randomUUID(), member.id, "borrow_confirmation", "Book borrowed", "You borrowed \"" + book.title + "\".", false, record.id, record.dueDate, TimeUtil.now()));

    HttpResponses.sendJson(exchange, 201, "{\"status\":\"success\",\"data\":{\"borrowRecord\":" + views.record(record) + "}}");
  }

  private User resolveIssueMember(HttpExchange exchange, String requestBody) throws IOException {
    String memberId = JsonUtil.jsonString(requestBody, "memberId", "");
    if (!memberId.isBlank()) {
      User member = store.findUser(memberId);
      if (member == null || !"member".equals(member.role)) {
        HttpResponses.sendError(exchange, 404, "Member not found.");
        return null;
      }
      return member;
    }

    String studentId = JsonUtil.jsonString(requestBody, "studentId", "").trim().toUpperCase(Locale.ROOT);
    String memberName = JsonUtil.jsonString(requestBody, "memberName", "").trim();
    String email = JsonUtil.jsonString(requestBody, "email", "").trim().toLowerCase(Locale.ROOT);

    if (studentId.isBlank() || memberName.isBlank()) {
      HttpResponses.sendError(exchange, 400, "Student name and student ID are required to issue books.");
      return null;
    }

    User existingByStudentId = store.findMemberByStudentId(studentId);
    if (existingByStudentId != null) {
      return existingByStudentId;
    }

    if (!email.isBlank()) {
      User existingByEmail = store.findUserByEmail(email);
      if (existingByEmail != null) {
        if (!"member".equals(existingByEmail.role)) {
          HttpResponses.sendError(exchange, 409, "Email is already used by another account.");
          return null;
        }
        return existingByEmail;
      }
    } else {
      email = studentId.toLowerCase(Locale.ROOT) + "@stackshelf.local";
    }

    User member = new User(
      "user-" + UUID.randomUUID(),
      memberName,
      email,
      "member",
      JsonUtil.jsonString(requestBody, "phone", ""),
      studentId,
      JsonUtil.jsonInt(requestBody, "semester", 1),
      JsonUtil.jsonInt(requestBody, "enrollmentYear", 2026),
      5,
      TimeUtil.now()
    );
    store.users().add(member);
    store.notifications().add(new Notification("notification-" + UUID.randomUUID(), member.id, "welcome", "Welcome to StackShelf", "Your member account was created at the issue desk.", false, "", null, TimeUtil.now()));
    return member;
  }

  void returnBook(HttpExchange exchange, String recordId) throws IOException {
    User user = security.requireAuth(exchange);
    if (user == null) {
      return;
    }

    BorrowRecord record = store.findBorrowRecord(recordId);
    if (record == null) {
      HttpResponses.sendError(exchange, 404, "Borrow record not found.");
      return;
    }
    if ("member".equals(user.role) && !record.memberId.equals(user.id)) {
      HttpResponses.sendError(exchange, 403, "Members can only return their own borrowed books.");
      return;
    }
    if ("returned".equals(record.status)) {
      HttpResponses.sendError(exchange, 400, "This book has already been returned.");
      return;
    }

    Book book = store.findBook(record.bookId);
    record.status = "returned";
    record.returnDate = TimeUtil.now();
    record.returnedToId = "librarian".equals(user.role) ? user.id : "";
    record.updatedAt = TimeUtil.now();
    if (book != null) {
      book.availableCopies += 1;
      book.updatedAt = TimeUtil.now();
    }
    store.notifications().add(new Notification("notification-" + UUID.randomUUID(), record.memberId, "return_confirmation", "Book returned", "You returned \"" + (book == null ? "the book" : book.title) + "\".", false, record.id, null, TimeUtil.now()));

    HttpResponses.sendJson(exchange, 200, "{\"status\":\"success\",\"data\":{\"borrowRecord\":" + views.record(record) + "}}");
  }

  void history(HttpExchange exchange) throws IOException {
    User user = security.requireAuth(exchange);
    if (user == null) {
      return;
    }

    borrowService.markOverdueRecords();
    Map<String, String> query = RequestUtil.query(exchange);
    String status = query.getOrDefault("status", "");
    List<BorrowRecord> rows = new ArrayList<>(store.borrowRecords());
    if ("member".equals(user.role)) {
      rows = rows.stream().filter(record -> record.memberId.equals(user.id)).collect(Collectors.toList());
    }
    if (!status.isBlank()) {
      rows = rows.stream().filter(record -> record.status.equals(status)).collect(Collectors.toList());
    }
    rows.sort(Comparator.comparing((BorrowRecord record) -> record.createdAt).reversed());
    int page = Math.max(1, TextUtil.parseInt(query.getOrDefault("page", "1"), 1));
    int limit = Math.min(100, Math.max(1, TextUtil.parseInt(query.getOrDefault("limit", "20"), 20)));
    sendRecords(exchange, rows, page, limit);
  }

  void overdue(HttpExchange exchange) throws IOException {
    User user = security.requireAuth(exchange);
    if (user == null || !security.requireLibrarian(exchange, user)) {
      return;
    }
    borrowService.markOverdueRecords();
    List<BorrowRecord> rows = store.borrowRecords().stream().filter(record -> "overdue".equals(record.status)).collect(Collectors.toList());
    HttpResponses.sendJson(exchange, 200, "{\"status\":\"success\",\"results\":" + rows.size() + ",\"data\":{\"records\":" + JsonUtil.array(rows, views::record) + "}}");
  }

  void availability(HttpExchange exchange, String bookId) throws IOException {
    User user = security.requireAuth(exchange);
    if (user == null) {
      return;
    }
    Book book = store.findBook(bookId);
    if (book == null) {
      HttpResponses.sendError(exchange, 404, "Book not found.");
      return;
    }
    HttpResponses.sendJson(exchange, 200, "{\"status\":\"success\",\"data\":{\"bookId\":" + JsonUtil.json(book.id) + ",\"title\":" + JsonUtil.json(book.title) + ",\"totalCopies\":" + book.totalCopies + ",\"availableCopies\":" + book.availableCopies + ",\"isAvailable\":" + (book.availableCopies > 0) + "}}");
  }

  private void sendRecords(HttpExchange exchange, List<BorrowRecord> rows, int page, int limit) throws IOException {
    int total = rows.size();
    int pages = Math.max(1, (int) Math.ceil(total / (double) limit));
    int start = Math.min(total, (page - 1) * limit);
    int end = Math.min(total, start + limit);
    List<BorrowRecord> pageRows = rows.subList(start, end);
    String pagination = JsonUtil.pagination(page, limit, total, pages);
    HttpResponses.sendJson(exchange, 200, "{\"status\":\"success\",\"results\":" + pageRows.size() + ",\"pagination\":" + pagination + ",\"data\":{\"records\":" + JsonUtil.array(pageRows, views::record) + "}}");
  }
}
